### DISCUSSION TITLE
Serve presigned URL from flask/fastapi/litestar

### DISCUSSION THREAD

**Discussion by jayceslesar** [minio/minio#19974] [Q&A] [answered]
Hey all, I have been banging my head against a wall trying to serve a generated presigned URL through an API framework in python.

I have a MinIO service and a Litestar service (akin to FastAPI or Flask) that uses the `Minio` python client to generate a signed URL for an object in blob storage. This URL comes back as `http://minio:9000/sign-stuff` which is not ideal because I have MinIO running on `http://localhost:9000`. Note that the API I am running is behind a domain of "my-domain.com", and that is all users of the application have access to.

I get a signature mismatch when I return that URL to the client (I replace minio with localhost) which makes sense because the host has changed.

What can I do to prevent this?

I am building the client with the URL of `minio:9000` as that is what its running as under docker, and the API container can hit that and interact with it fine.

NginX is not running in a container and the configuration is as follows:
```
server {
    listen 443 ssl;
    server_name my-domain.com;

    ssl_certificate /etc/ssl/cert.pem;
    ssl_certificate_key /etc/ssl/key.pem;

    ssl_client_certificate /etc/ssl/cloudflare.crt;
    ssl_verify_client on;

    ignore_invalid_headers off;

    location / {
        proxy_pass http://127.0.0.1:8000;

        # Include headers to preserve the original client's IP address
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
	proxy_ssl_server_name on;

        # timeout values
        proxy_connect_timeout       600s;
        proxy_send_timeout          600s;
        proxy_read_timeout          600s;

        # ws support
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
    }


    # minio proxy
    location /demoblobs {
        proxy_pass http://127.0.0.1:9000;
        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_connect_timeout 300;
        proxy_http_version 1.1;
        proxy_set_header Connection "";
        chunked_transfer_encoding off;
    }
}
```
Here is the `MinIO` configuration in the compose: https://github.com/MegaAntiCheat/masterbase/blob/main/docker-compose.yaml#L3

What can I possibly be doing wrong? Every other example I can find online is slightly different or half baked.

**Comment by klauspost** [comment, marked as answer]
You cannot serve S3 on a subpath.

**Reply by jayceslesar** [reply]
Thanks for the info, I could serve it out of a subdomain though correct?

**Reply by jayceslesar** [reply]
And even though im using my own blob storage on my server I cant serve on a subpath? (No AWS in the loop anywhere in my workflow)

**Reply by klauspost** [reply]
Yes. Subdomain is fine.

You can't change the URL in a pre-signed request - then it isn't the resource you are signing any more.

**Reply by jayceslesar** [reply]
Thank you! I suppose just reading the contents from minio and serving as bytes response to the client will also work fine and signed URLs will need a subdomain.

**Reply by klauspost** [reply]
Regular signatures should also match. Subpaths are just a bad idea.
